oioswift/common/middleware/container_hierarchy.py

```diff
@@ -43,6 +44,106 @@
+            -- Get index of the first delimiter of the key
+            local index = string.find(key, delimiter, start)
+            if index  == nil then
+                return #key
+            end
+            return index
+        end
+        local get_first_level = function(key, start, delimiter)
+            -- Get the hierarchy first level
+            -- start = start looking at this index
+            -- delimiter = used for separating level
+            -- key = prefix .. delimiter .. first_level ...
+            -- Examples get_first_level("aaa/bbb/ccc/", 4, '/') = "aaa/bbb/"
+            -- get_first_level("aaa/bbb/ccc/", 8, '/') = "aaa/bbb/ccc/"
+
+            local index = get_index(key, start, delimiter)
+            local rv = string.sub(key, 0, index)
+            return rv
+        end
+
+        local bucket = KEYS[1]
+        local prefix = KEYS[2]
+        local delimiter = KEYS[3]
+        local marker = KEYS[4]
+        local recursive = KEYS[5]
+        local limit = tonumber(KEYS[6])
+        local prefix_len = string.len(prefix)
+        local set = {}
+        -- There is no set on lua this is a hack using the table keys to hold
+        -- data and assure uniqueness
+    """
+
+    lua_script_zkeys = lua_script_utilities + """
+         -- getn() doesn't count non numerical key
+        local get_size = function(table)
+            local count = 0
+            for _ in pairs(table) do
+                count = count + 1
+            end
+            return count
+        end
+
+        local keys = {}
+        --rank is the redis term for zset index https://redis.io/commands/zrank
+        local rank = 0
+        local finish = false
+
+        if marker == "" then
+            marker = "-"
+        else
+            marker = "[" .. marker
+        end
+
+        local nb_elem = redis.call('ZCOUNT', bucket, '-inf', '+inf')
+        local count = 0
+        while count < limit + 2  and finish == false do
+            local list_until = math.min(rank + limit + 2, nb_elem)
+            keys = redis.call('ZRANGEBYLEX', bucket, marker, '+', 'LIMIT',
+                               0, limit)
+            redis.log(redis.LOG_WARNING, " keys = ", keys)
+            for i=1, #keys do
+                 redis.log(redis.LOG_WARNING, " key = ", keys[i])
+                 local elem = keys[i]
+                 if recursive ~= '1' then
+                     elem = get_first_level(keys[i], prefix_len + 1,
+                                            delimiter)
+                 end
+
+                 local check = true
+                 if prefix ~= "" then
+                     local index = string.find(elem, prefix)
+                     if index == nil or index > 1 then
+                         check = false
+                     end
+                 end
+
+                 if check and set[elem] == nil then
+                     count = count + 1
+                     set[elem] = true
+                 end
```

Comment:
We can imagine an optimization here in a dedicated PR:
If we have lot of keys like below and we asked a non-recursive listing on /
- dir1/dir2/...
- dir1/dir3/...
...
- dir1/dir100000/...
- dir2/...
we could stop the loop and restart maker with "[" .. "dir1/" + \xef\xbf\xbd"
(like https://github.com/open-io/oio-sds/pull/1689/files#diff-3b3bc99683f0c5fe810b2291cc9beeafR1518)